### PR TITLE
feat(trajectory_follower): publsih control horzion

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

It was previously agreed that autoware_msgs should include control horizon. However, until now the msgs has not been used.
https://github.com/autowarefoundation/autoware_msgs/pull/11

In this PR, publish control horizon topic as experimental one. 
Although this is an informal interface, it is a useful topic for users who want to maximize control performance in vehicular nodes.

If the longitudinal and lateral time steps are different, calculate the greatest common divisor time step and align the sizes with zero-order hold.
If one size is 1, repeat to match the other size.

image

## Related links

https://github.com/autowarefoundation/autoware_msgs/pull/11

**Parent Issue:**

- Link

internal slack: https://star4.slack.com/archives/C4P0NSMB5/p1731572458828639

## How was this PR tested?

- psim
- unit test
- 2024/11/14 https://evaluation.tier4.jp/evaluation/reports/b99d67ea-b495-5b42-a579-862912e27283/?project_id=prd_jt

## Notes for reviewers

None.

## Interface changes

None.

## Effects on system behavior

None.

### CODE REVIEW

**Comment by kosuke55** [comment]
@TakaHoribe 
Could you please check briefly

- it looks good in general?
- the topic name “experimental” is appropriate or any other good name?

If the general policy looks good, I will add tests and documents.

File: control/autoware_mpc_lateral_controller/src/mpc_lateral_controller.cpp [hunk 1]
```diff
@@ -304,9 +306,13 @@ trajectory_follower::LateralOutput MpcLateralController::run(
   publishPredictedTraj(predicted_traj);
   publishDebugValues(debug_values);
 
-  const auto createLateralOutput = [this](const auto & cmd, const bool is_mpc_solved) {
+  const auto createLateralOutput =
+    [this](
+      const auto & cmd, const bool is_mpc_solved,
+      const auto & cmd_horion) -> trajectory_follower::LateralOutput {
```
Comment (TakaHoribe) [inline control/autoware_mpc_lateral_controller/src/mpc_lateral_controller.cpp:312]: Fix typos.

File: control/autoware_mpc_lateral_controller/test/test_mpc.cpp [hunk 2]
```diff
@@ -241,8 +245,9 @@ TEST_F(MPCTest, InitializeAndCalculateRightTurn)
   Lateral ctrl_cmd;
   Trajectory pred_traj;
   Float32MultiArrayStamped diag;
+  LateralHorizon ctrl_cmd_horizon;
   const auto odom = makeOdometry(pose_zero, default_velocity);
-  ASSERT_TRUE(mpc->calculateMPC(neutral_steer, odom, ctrl_cmd, pred_traj, diag));
+  ASSERT_TRUE(mpc->calculateMPC(neutral_steer, odom, ctrl_cmd, pred_traj, diag, ctrl_cmd_horizon));
```
Comment (TakaHoribe) [inline control/autoware_mpc_lateral_controller/test/test_mpc.cpp:250]: Please add tests for `horizon` as `ctrl_cmd`.

```
  EXPECT_LT(ctrl_cmd.steering_tire_angle, 0.0f);
  EXPECT_LT(ctrl_cmd.steering_tire_rotation_rate, 0.0f);
```

File: control/autoware_trajectory_follower_node/src/controller_node.cpp [hunk 3]
```diff
@@ -254,6 +286,75 @@ void Controller::publishProcessingTime(
   pub->publish(msg);
 }
 
+std::optional<ControlHorizon> Controller::createControlHorizon(
```
Comment (TakaHoribe) [inline control/autoware_trajectory_follower_node/src/controller_node.cpp:289]: The function name `createControlHorizon` does not clearly convey this functionality. It is better to include terms such as mergeLatLonHorizon or Resample to clarify its purpose.

**Review by TakaHoribe — COMMENTED** [review]
Overall lgtm. Please ask @takayuki5168 to check the details.

**Comment by TakaHoribe** [comment]
@kosuke55 And please add `enable` flag if it is experimental.

File: control/autoware_mpc_lateral_controller/src/mpc_lateral_controller.cpp (hunk 1 above)
Comment (kosuke55) [reply control/autoware_mpc_lateral_controller/src/mpc_lateral_controller.cpp]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/8977/commits/3aea463c6b04a4bd911af38e16e0000ebe66861d

**Comment by kosuke55** [comment]
>  And please add enable flag if it is experimental.

added in https://github.com/autowarefoundation/autoware.universe/pull/8977/commits/d6bf00099d273c86320fe89645a822ddb5c049c6

File: control/autoware_mpc_lateral_controller/test/test_mpc.cpp (hunk 2 above)
Comment (kosuke55) [reply control/autoware_mpc_lateral_controller/test/test_mpc.cpp:253]: added tests for horizon in 
a53f6d0

**Pull request merged**